### PR TITLE
Fix for white hairlines rendering in pdf

### PR DESCRIPTION
**Pull request by brianholleibm**
This fix will provide a solution to the following issues:

#3351
#4860
#3331
#9464 (Closed) (Duplicate)
#9481 (Closed) (Duplicate)
#9829 (Closed) (Duplicate)
#9953 (Closed) (Duplicate)

This solution takes into account the resolution of the viewer before painting images to the canvas and rounds the painting to the nearest full pixel. The problem in the issues above is caused by drawing to a canvas with a width/height that is more accurate than pixel width allows. This results in the drawing coloring part of a pixel, which causes a partial transparency. This transparency creates what appears to be a thin line between images. Rounding the images to the nearest pixel insures there will be no partially filled pixels, and therefore no thin lines.

Performance Tests
```
Run1:
browser | stat         | Count | Baseline(ms) | Current(ms) |  +/- |     %  | Result(P<.05)
------- | ------------ | ----- | ------------ | ----------- | ---- | ------ | -------------
Firefox | Overall      |    50 |        17427 |       16954 | -474 |  -2.72 |              
Firefox | Page Request |    50 |           11 |           9 |   -2 | -14.34 |              
Firefox | Rendering    |    50 |        17416 |       16944 | -472 |  -2.71 |   
```
```
Run2:
browser | stat         | Count | Baseline(ms) | Current(ms) |   +/- |     %  | Result(P<.05)
------- | ------------ | ----- | ------------ | ----------- | ----- | ------ | -------------
Firefox | Overall      |    50 |        18640 |       15158 | -3481 | -18.68 |        faster
Firefox | Page Request |    50 |           11 |          10 |    -1 |  -5.59 |              
Firefox | Rendering    |    50 |        18629 |       15148 | -3481 | -18.69 |        faster
```

### CODE REVIEW

File: src/display/canvas.js
Comment (Snuffleupagus) [inline src/display/canvas.js]: Having these be defined as
```javascript
origX = args[j++];
origY = args[j++];
```
and then simply using `origX`/`origY` when defining `x`/`y` just below instead seems more readable (rather than having a bunch of extra `args[j++]` occurrences).

File: src/display/canvas.js
Comment (Snuffleupagus) [inline src/display/canvas.js]: Please remember to use `let`/`const` rather than `var` when adding *new* code.

File: src/display/canvas.js
Comment (Snuffleupagus) [inline src/display/canvas.js]: Nit: It might perhaps be a tiny bit more efficient to avoid extra function calls here, and in the `width === 0` case just above, by utilizing the now already available `pixelSize` value instead.

File: src/display/canvas.js
Comment (ThomasdenH) [reply src/display/canvas.js]: Because of the loop, `args[j++]` is not the same as `args[0]`.

File: src/display/canvas.js
Comment (brianholle) [reply src/display/canvas.js]: I am new to html canvas's but here was my logic. Please verify I am looking at this correctly.
The goal is to extend all images outward to the nearest pixel. Compressing inward would generate gaps between images.

With each (x,y) point represented in the canvas with (0,0) being the top left most point.

Here, we are drawing a rectangle. So the goal is to extend each point as follows:

Top Left (x,y):  x rounds towards left (floor). Y rounds north towards y=0 (ceil)
Top Right (xw, y): xw rounds towards right (ceil). Y rounds north (ceil).
Bottom Right (xw, yh): xw rounds towards right (ceil). yh rounds south (while represented by positive value(?), so, to make it as large as possible, ceil.
Bottom Left (x, yh): x rounds towards left (floor). yh rounds south (ceil).

I have some confusion over how the Y variable is treated here. Is its value negative, considering its under the y=0 line (top of document)? Var height is set to negative in my example cases, while var Y is set to positive. Var YH is defined as height+y, where Y is normally always larger than height. This confusion could possibly lead to incorrect rounding.

File: src/display/canvas.js
Comment (timvandermeij) [inline src/display/canvas.js]: `var` should be `const`.

File: src/display/canvas.js
Comment (timvandermeij) [inline src/display/canvas.js]: `var` should be `const`.

**Review by timvandermeij — COMMENTED** [review]
After making these changes, could you squash the commits into one? Refer to https://github.com/mozilla/pdf.js/wiki/Squashing-Commits for how to do that easily if you're not familiar with the process. It should get rid of the merge commit too.

File: src/display/canvas.js
Comment (brianholleibm) [reply src/display/canvas.js]: @timvandermeij  If you have a moment, please review this comment. I would like a second opinion before we merge, it may be possible that I have made a mistake in my decision to use .ceil / .floor in specific cases.

**Review by brendandahl — CHANGES_REQUESTED** [review]
Looks like there a few real regressions where lines either disappear or are now too bold. Can you look into the failures?

File: src/display/canvas.js
Comment (brianholle) [reply src/display/canvas.js]: Actually, it seems I get better results with the following:
```
x = Math.floor(origX / pixelSize) * pixelSize;
y = Math.floor(origY / pixelSize) * pixelSize;

xw = Math.ceil((origX + width) / pixelSize) * pixelSize;
yh = Math.ceil((origY + height) / pixelSize) * pixelSize;
```
I believe that the canvas actually treats Y height as positive, with y=0 being the top of the page and y>0+n = bottom.

Where my confusion comes from, I think, is the example file that I am using. When drawing rectangles from the top of the page to the bottom, the following values are being passed into "constructPath": (571.41, 684.211, 5.303, -578.104)

These values are X, Y, Width, Height respectively. What is strange here is that, according to my believe of X/Y values on a canvas, (X,Y) actually represents the lower left hand corner of our rectangle instead of the expected upper left corner in this example. 

When YH is calculated, we are adding the negative height to our Y value, giving us the top side of our rectangle. In this case, we would want to round this value towards the top of the page (y=0) so we would want to floor this value.

However, if the usual case is that (X,Y) is the upper left corner of our image (like I originally expected), that would mean that height would be a positive value, and therefore YH would represent the bottom of our rectangle. In this case, we would want to make YH rounded upward towards the bottom of the document, aka Math.ceil(YH).

I am not too sure what determines the values of (X,Y, Height, Width) when passed into "constructPath." Can an assumption ever be made about which point (top left, top right, bottom left, bottom right) (X,Y) actually represents? If not, I need to add more code to determine the location of (X,Y) before making assumptions about the direction of rounding.

Comment (brianholle) [reply src/display/canvas.js]: @brendandahl

**Comment by marktanner1331** [comment]
The issue is unfortunately more complex than believed:

A majority of PDF's will first set up a clipping path before drawing the image. The clipping path will be the exact dimensions of the image. 

This clipping path will suffer from the same sub-pixel rendering issues as described above, and cause the white lines to appear even though the image is technically the right size. 

To test this theory, try commenting out the clipping code in canvas.consumePath() leaving only the ctx.beginPath(); you'll notice that the white lines are gone.

**Comment by ssaring** [comment]
We have the same rendering problem in our current project, it's reported in issue #11400.
It would be very helpful for us to know whether there's a chance for solving these problems.

Unfortunately I have no programming skills in the PDF and rendering context, is there any other way to support the progress?

**Comment by jasonbolbar** [comment]
Is there any update for this? I'm interested in the solution for this issue. I tried to check the code cloning the repo from @brianholleibm and checking out to **3351_White_Bar_Fix** branch, later I ran **gulp generic** in order to generate the viewer and finally I tested in an iPhone using BrowserStack getting this result:
phone_test

**Comment by Snuffleupagus** [comment]
Closing for now, since there's not been any updates for a couple of years and as-is this patch cannot land given that there's too many regressions in existing test-cases; a number of issues remain open to track this (please see https://github.com/mozilla/pdf.js/pull/10016#issue-211238996 for a list).

**Pull request closed without merging**
